tests/subsys/net/lib/nrf_cloud/cloud/src/main.c

```diff
@@ -0,0 +1,1880 @@
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud successfully with dc_connect */
+	connect_cloud_dc_success();
+
+	nct_dc_stream_fake.custom_fake = fake_nct_dc_stream__succeeds;
+	nct_dc_send_fake.custom_fake = fake_nct_dc_send__succeeds;
+
+	/* Message data */
+	struct nrf_cloud_tx_data tx_data_t = {
+		.topic_type = NRF_CLOUD_TOPIC_MESSAGE,
+		.qos = MQTT_QOS_0_AT_MOST_ONCE
+	};
+
+	int ret = nrf_cloud_send(&tx_data_t);
+
+	zassert_ok(ret, "return should be 0 when success");
+
+	tx_data_t.qos = MQTT_QOS_1_AT_LEAST_ONCE;
+	ret = nrf_cloud_send(&tx_data_t);
+	zassert_ok(ret, "return should be 0 when success");
+}
+
+/* Verify that the function will fail if the cloud is
+ * in dc_connected state with the topic_type is equal
+ * topic_bulk and nct_dc_bulk_send fails
+ */
+ZTEST(nrf_cloud_test, test_cloud_send_topic_bulk_dc_bulk_send_failed)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud successfully with dc_connect */
+	connect_cloud_dc_success();
+
+	nct_dc_bulk_send_fake.custom_fake = fake_nct_dc_bulk_send__fails;
+
+	/* Message data */
+	struct nrf_cloud_tx_data tx_data_t = {
+		.topic_type = NRF_CLOUD_TOPIC_BULK,
+		.qos = MQTT_QOS_0_AT_MOST_ONCE
+	};
+
+	int ret = nrf_cloud_send(&tx_data_t);
+
+	zassert_not_equal(0, ret, "return should not be 0 when nct_dc_bulk_send failed");
+}
+
+/* Verify that the function will succeed if the cloud is
+ * in dc_connected state with the topic_type is equal
+ * topic_bulk and nct_dc_bulk_send succeeds
+ */
+ZTEST(nrf_cloud_test, test_cloud_send_topic_bulk_success)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud successfully with dc_connect */
+	connect_cloud_dc_success();
+
+	nct_dc_bulk_send_fake.custom_fake = fake_nct_dc_bulk_send__succeeds;
+
+	/* Message data */
+	struct nrf_cloud_tx_data tx_data_t = {
+		.topic_type = NRF_CLOUD_TOPIC_BULK,
+		.qos = MQTT_QOS_0_AT_MOST_ONCE
+	};
+
+	int ret = nrf_cloud_send(&tx_data_t);
+
+	zassert_ok(ret, "return should be 0 when success");
+}
+
+/* Verify that the function will fail if an invalid topic_type is given */
+ZTEST(nrf_cloud_test, test_cloud_send_invalid_topic_type)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Message data */
+	struct nrf_cloud_tx_data tx_data_t = {
+		.topic_type = 0,
+		.qos = MQTT_QOS_0_AT_MOST_ONCE
+	};
+
+	int ret = nrf_cloud_send(&tx_data_t);
+
+	zassert_equal(-ENODATA, ret, "return should be -ENODATA with invalid topic type");
+}
+
+#if defined(CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD)
+
+/* TEST START_CONNECTION_POLL */
```

Comment:
This one could go in either the MQTT-mocked or nct/fsm-mocked suites.

I would personally probably lean towards the mocked MQTT suite since that will do a better job of testing how the connection poll thread handles the asynchronicity introduced by MQTT.

Reply:
BTW, a convenient "boundary of separation" for the two suites would be whether or not they involve transmission. nct/fsm-mocked suite for tests that are non-transmit, non-receive, and MQTT-mocked tests that do transmit and/or receive